### PR TITLE
Add node controller

### PR DESCRIPTION
**Pull request by r7vme**
node-controller pod will be running in the same namespace as guest cluster pods. This change just adds new deployment.

Please approve if ok, i'll update vendor as soon as kvmtpr will be merged.

Additional info:
node-controller responsible for managing (deleting) dead nodes from guest cluster based on host cluster state.

Part of: giantswarm/k8scloudconfig#162

All PRs (all tested together):
https://github.com/giantswarm/kvm-operator-node-controller/pull/4
https://github.com/giantswarm/kubernetesd/pull/289
https://github.com/giantswarm/kvm-operator/pull/221
https://github.com/giantswarm/kvmtpr/pull/51

### CODE REVIEW

File: service/key/key.go [hunk 1]
```diff
@@ -12,10 +12,15 @@ import (
 )
 
 const (
-	MasterID = "master"
-	WorkerID = "worker"
+	MasterID         = "master"
+	WorkerID         = "worker"
+	NodeControllerID = "node-controller"
```
Comment (xh3b4sd) [inline service/key/key.go]: Please keep the alphabetical order.

File: service/resource/deployment/resource_test.go [hunk 2]
```diff
@@ -21,6 +21,8 @@ import (
 	extensionsv1 "k8s.io/client-go/pkg/apis/extensions/v1beta1"
 )
 
+const ExpectedNodeCtrlCount = 1
```
Comment (xh3b4sd) [inline service/resource/deployment/resource_test.go]: Please make this part of the test case. There are no package wide constants for the master and worker counts. So lets keep it within the test spec.

File: service/key/key.go (hunk 1 above)
Comment (r7vme) [reply service/key/key.go]: ok

File: service/resource/deployment/resource_test.go (hunk 2 above)
Comment (r7vme) [reply service/resource/deployment/resource_test.go]: ok

**Comment by r7vme** [comment]
@xh3b4sd addressed your comments.

File: service/resource/deployment/resource.go [hunk 3]
```diff
@@ -121,6 +121,14 @@ func (r *Resource) GetDesiredState(ctx context.Context, obj interface{}) (interf
...
 		}
 		deployments = append(deployments, workerDeployments...)
+
+		if key.HasNodeController(customObject) {
+			nodeControllerDeployment, err := newNodeControllerDeployment(customObject)
+			if err != nil {
+				return nil, microerror.Mask(err)
+			}
+			deployments = append(deployments, nodeControllerDeployment)
+		}
```
Comment (xh3b4sd) [inline service/resource/deployment/resource.go]: FYI IMO this is a hack we will stop doing quite soon. The implication of adding such functionality will soon be to create a new version bundle to create a new minor release of the kvm-operator. We would then copy the deployment resource (or all maybe all?) and route custom objects within the operatorkit framework accordingly. Lets please create an issue to track the cleanup of this code. At some point all custom objects should have the node controller configuration and we can remove the condition.

File: service/key/key.go [hunk 4]
```diff
@@ -70,6 +76,13 @@ func EtcdPVCName(clusterID string, vmNumber string) string {
 	return fmt.Sprintf("%s-%s-%s", "pvc-master-etcd", clusterID, vmNumber)
 }
 
+func HasNodeController(customObject kvmtpr.CustomObject) bool {
+	if customObject.Spec.KVM.NodeController != (kvmtprkvm.NodeController{}) {
+		return true
+	}
+	return false
+}
```
Comment (xh3b4sd) [inline service/key/key.go]: This detection is nice. I like it.

File: service/resource/deployment/resource.go (hunk 3 above)
Comment (r7vme) [reply service/resource/deployment/resource.go]: Here we go https://github.com/giantswarm/kvm-operator/issues/223

**Comment by r7vme** [comment]
Great, thx for review!!!

**Pull request merged**